packages/expo-in-app-purchases/android/src/main/java/expo/modules/inapppurchases/BillingManager.java

```diff
@@ -0,0 +1,530 @@
+                  promise.resolve(response);
+                }
+              });
+            }
+          });
+      }
+    };
+
+    executeServiceRequest(queryToExecute);
+  }
+
+  /**
+   * Format the result of a Purchase or Sku Details query depending on the result code
+   */
+  public static Bundle formatResponse(BillingResult billingResult, ArrayList<? extends Parcelable> results) {
+    Bundle response = new Bundle();
+    int responseCode = billingResult.getResponseCode();
+    if (responseCode == BillingResponseCode.OK) {
+      response.putInt("responseCode", OK);
+      response.putParcelableArrayList("results", results != null ? results : new ArrayList<Parcelable>());
+    } else if (responseCode == BillingResponseCode.USER_CANCELED) {
+      response.putInt("responseCode", USER_CANCELED);
+    } else {
+      response.putInt("responseCode", ERROR);
+      response.putInt("errorCode", errorCodeNativeToJS(responseCode));
+    }
+    return response;
+  }
+
+  /**
+   * Convert native error code to match corresponding TS enum
+   */
+  private static int errorCodeNativeToJS(int responseCode) {
+    switch (responseCode) {
+      case BillingResponseCode.ERROR:
+        return 0;
+      case BillingResponseCode.FEATURE_NOT_SUPPORTED:
+        return 1;
+      case BillingResponseCode.SERVICE_DISCONNECTED:
+        return 2;
+      case BillingResponseCode.SERVICE_UNAVAILABLE:
+        return 3;
+      case BillingResponseCode.SERVICE_TIMEOUT:
+        return 4;
+      case BillingResponseCode.BILLING_UNAVAILABLE:
+        return 5;
+      case BillingResponseCode.ITEM_UNAVAILABLE:
+        return 6;
+      case BillingResponseCode.DEVELOPER_ERROR:
+        return 7;
+      case BillingResponseCode.ITEM_ALREADY_OWNED:
+        return 8;
+      case BillingResponseCode.ITEM_NOT_OWNED:
+        return 9;
+    }
+    return 0;
+  }
+
+  /**
+   * Convert native purchase state to match corresponding TS enum
+   */
+  private static int purchaseStateNativeToJS(int purchaseState) {
+    switch (purchaseState) {
+      case Purchase.PurchaseState.PENDING:
+        return 0;
+      case Purchase.PurchaseState.PURCHASED:
+        return 1;
+      case Purchase.PurchaseState.UNSPECIFIED_STATE:
+        return 2;
+    }
+    return 0;
+
+  }
+
+  private static Bundle skuToBundle(SkuDetails skuDetails) {
+    Bundle bundle = new Bundle();
+
+    String subscriptionPeriod =
+      skuDetails.getType().equals(SkuType.SUBS) ?
+        skuDetails.getSubscriptionPeriod() :
+        INAPP_SUB_PERIOD;
+
+    bundle.putString("description", skuDetails.getDescription());
+    bundle.putString("price", skuDetails.getPrice());
+    bundle.putLong("priceAmountMicros", skuDetails.getPriceAmountMicros());
+    bundle.putString("priceCurrencyCode", skuDetails.getPriceCurrencyCode());
+    bundle.putString("productId", skuDetails.getSku());
+    bundle.putString("title", skuDetails.getTitle());
+    bundle.putString("type", skuDetails.getType());
```

Comment:
Convert `skuDetails.getType()` to a TS-defined value. The intent is to provide a universal API, not to call the Android API from JS.

Reply:
Resolved in https://github.com/expo/expo/pull/4416/commits/9914e420a5e599bdbf1b300a3508e19dd9cb5265